encoder/flatfile.go

```diff
@@ -0,0 +1,225 @@
+package encoder
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	common_datalayer "github.com/mimiro-io/common-datalayer"
+	"io"
+	"strings"
+)
+
+func NewFlatFileItemFactory() ItemFactory {
+	return &FlatFileItemFactory{}
+}
+
+type FlatFileItemFactory struct{}
+
+func (c *FlatFileItemFactory) NewItem() common_datalayer.Item {
+	return &FlatFileItem{data: make(map[string]any)}
+}
+
+type FlatFileItemWriter struct {
+	writer    io.WriteCloser
+	batchInfo *common_datalayer.BatchInfo
+	fields    []FlatFileField
+}
+
+// use position of field in the list to order fieldOrder
+type FlatFileField struct {
+	Name   string `json:"name"`
+	Length int    `json:"length"`
+	Ignore bool   `json:"ignore"`
+	Type   string `json:"type"`
+}
+
+func NewFlatFileItemWriter(sourceConfig map[string]any, data io.WriteCloser, batchInfo *common_datalayer.BatchInfo) (*FlatFileItemWriter, error) {
+	writer := &FlatFileItemWriter{writer: data, batchInfo: batchInfo}
+	// do this in function or here?
+	itemWriter, err2 := NewFlatFileWriteConfig(sourceConfig["fields"].([]map[string]interface{}), writer)
+	if err2 != nil {
+		return itemWriter, err2
+	}
+	return writer, nil
+}
+
+func NewFlatFileWriteConfig(fields []map[string]interface{}, writer *FlatFileItemWriter) (*FlatFileItemWriter, error) {
+	var config []FlatFileField
+	jsonData, err := json.Marshal(fields)
+	err = json.Unmarshal(jsonData, &config)
+	if err != nil {
+		fmt.Println("Error:", err)
+		return nil, err
+	}
+	if config == nil {
+		return nil, fmt.Errorf("missing field config for flat file")
+	} else {
+		writer.fields = config
+	}
+	return nil, nil
+}
+
+func NewFlatFileReadConfig(fields []map[string]interface{}, reader *FlatFileItemIterator) (*FlatFileItemIterator, error) {
+	var config []FlatFileField
+	jsonData, err := json.Marshal(fields)
+	err = json.Unmarshal(jsonData, &config)
+	if err != nil {
+		fmt.Println("Error:", err)
+		return nil, err
+	}
+	if config == nil {
+		return nil, fmt.Errorf("missing field config for flat file")
+	} else {
+		reader.fields = config
+	}
+	return nil, nil
+}
+func appendSpaces(value string, amount int) string {
+	for i := 0; i < amount; i++ {
+		value += " "
+	}
+	return value
+}
+
+func prependZeros(value string, amount int) string {
+	prefix := ""
+	for i := 0; i < amount; i++ {
+		prefix += "0"
+	}
+	return prefix + value
+}
+
+func (c *FlatFileItemWriter) Close() error {
+	return c.writer.Close()
+}
+func (c *FlatFileItemWriter) Write(item common_datalayer.Item) error {
+	buf := new(bytes.Buffer)
+	row := item.NativeItem().(map[string]any)
```

Comment:
while this is ok, it does expose the underlying impl. Suggest if possible to use the exposed functions connected to Item.